Assets/Scripts/Game/GameManager.cs

```diff
@@ -59,23 +61,43 @@ private IEnumerator HandleWallsDangerousness()
             {
                 yield return null;
 
-                while (!_waveManager.CurrentWave.switchOnWalls)
+                MoveRestingWallsToDangerous();
+
+                while (_waveManager.CurrentWave.dangerWallAmount == 0 || 
+                    _waveManager.CurrentWave.dangerWallAmount <= _dangerWallsCurrently ||
+                    _wallsToBeDangerous.Count == 0)
                 {
+                    MoveRestingWallsToDangerous();
                     yield return new WaitForSeconds(1);
                 }
 
-                var dangerousInterval = _wallDangerousInterval + _wallWarningInterval;
-
-                var wall = WallsToBeDangerous[new Random().Next(WallsToBeDangerous.Count)];
-                
-                StartCoroutine(wall.GetComponent<WallDanger>().BecameDangerousCoroutine(_wallWarningInterval));
+                var wall = _wallsToBeDangerous[new Random().Next(_wallsToBeDangerous.Count)];
+                _dangerWallsCurrently++;
+                wall.CanBeDangerous = false;
+                StartCoroutine(wall.BecameDangerousCoroutine(
+                    () => _dangerWallsCurrently--,
+                    _wallWarningInterval,
+                    _wallDangerousInterval,
+                    _wallCoolDownInterval));                
+                MoveWallToRestingWalls(wall);
+            }
+        }
 
-                yield return new WaitForSeconds(dangerousInterval);
+        // TODO Possibly should be implemented as independent coroutine 
+        // and removed from HandleWallsDangerousness()
+        private void MoveRestingWallsToDangerous()
+        {
+            if (_restingWalls.Count == 0) return;
 
-                wall.GetComponent<WallDanger>().BecameSafe();
+            var wallsToMove = _restingWalls.Where(w => w.CanBeDangerous).ToList();
+            _restingWalls = _restingWalls.Except(wallsToMove).ToList();
+            _wallsToBeDangerous.AddRange(wallsToMove);
+        }
 
-                yield return new WaitForSeconds(_wallCoolDownInterval);
-            }
+        private void MoveWallToRestingWalls(WallDanger wall)
```

Comment:
Please inline, actions are simple and self-explanatory.

Reply:
Ok